components/movementsensor/gpsrtk/pmtk.go

```diff
@@ -0,0 +1,197 @@
+	return makeRTKI2C(ctx, deps, conf, logger, nil)
+}
+
+// makeRTKI2C is separate from newRTKI2C, above, so we can pass in a non-nil mock I2C bus during
+// unit tests.
+func makeRTKI2C(
+	ctx context.Context,
+	_ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+	mockI2c buses.I2C,
+) (movementsensor.MovementSensor, error) {
+	newConf, err := resource.NativeConfig[*I2CConfig](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	g := &gpsrtk{
+		Named:      conf.ResourceName().AsNamed(),
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+		logger:     logger,
+		err:        movementsensor.NewLastError(1, 1),
+	}
+
+	ntripConfig := &gpsutils.NtripConfig{
+		NtripURL:             newConf.NtripURL,
+		NtripUser:            newConf.NtripUser,
+		NtripPass:            newConf.NtripPass,
+		NtripMountpoint:      newConf.NtripMountpoint,
+		NtripConnectAttempts: newConf.NtripConnectAttempts,
+	}
+
+	g.ntripClient, err = gpsutils.NewNtripInfo(ntripConfig, g.logger)
+	if err != nil {
+		return nil, err
+	}
+	g.InputProtocol = "i2c"
+
+	config := gpsutils.I2CConfig{
+		I2CBus:      newConf.I2CBus,
+		I2CBaudRate: newConf.I2CBaudRate,
+		I2CAddr:     newConf.I2CAddr,
+	}
+	if config.I2CBaudRate == 0 {
+		config.I2CBaudRate = 115200
+	}
+
+	// If we have a mock I2C bus, pass that in, too. If we don't, it'll be nil and constructing the
+	// reader will create a real I2C bus instead.
+	dev, err := gpsutils.NewI2cDataReader(config, mockI2c, logger)
+	if err != nil {
+		return nil, err
+	}
+	g.cachedData = gpsutils.NewCachedData(dev, logger)
+
+	g.correctionWriter, err = newCorrectionWriter(newConf.I2CBus, byte(newConf.I2CAddr))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := g.start(); err != nil {
+		return nil, err
+	}
+
+	// It's possible that we've taken so long to start up that the resource manager has given up on
+	// us and tried constructing a new component instead. If that happens, we don't want 2
+	// components talking to the same chip. So, if our context is canceled, close our component
+	// instead of returning it.
+	if ctx.Err() != nil {
+		logger.Warn("context canceled by the end of the constructor! Closing the new component...")
+		return nil, fmt.Errorf("timed out constructing I2C RTK reader. Closing: %w", g.Close(ctx))
+	}
+
+	return g, nil
+}
+
+func newCorrectionWriter(busname string, address byte) (io.ReadWriteCloser, error) {
+	bus, err := buses.NewI2cBus(busname)
+	if err != nil {
+		return nil, err
+	}
+	handle, err := bus.OpenHandle(address)
+	if err != nil {
+		return nil, err
+	}
+	correctionWriter := i2cCorrectionWriter{
+		bus:    bus,
+		handle: handle,
+	}
+	return &correctionWriter, nil
+}
+
+// This implements the io.ReadWriteCloser interface.
+type i2cCorrectionWriter struct {
+	bus    buses.I2C
+	handle buses.I2CHandle
+}
+
+func (i *i2cCorrectionWriter) Read(p []byte) (int, error) {
+	return 0, errors.New("unimplemented")
+}
```

Comment:
What's this for?

Reply:
`gpsutils.GetGGAMessage` requires an `io.ReadWriteCloser`, though @JohnN193 has plans to change it to just an `io.WriteCloser` in the near future, and then we can remove this again. I'll add in a comment.